bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/WatchdogInterceptor.java

```diff
@@ -0,0 +1,208 @@
+package com.google.cloud.bigtable.grpc.io;
+
+import com.google.api.core.ApiClock;
+import com.google.common.base.Preconditions;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.concurrent.GuardedBy;
+import org.joda.time.Duration;
+
+/**
+ * Internal implementation detail to prevent RPC from hanging
+ */
+public class WatchdogInterceptor implements ClientInterceptor {
+  // Dummy value to convert the ConcurrentHashMap into a Set
+  private static Object PRESENT = new Object();
+  private final ConcurrentHashMap<WatchedCall<?,?>, Object> openStreams = new ConcurrentHashMap<>();
+
+  private final Set<MethodDescriptor<?,?>> watchedMethods;
+  private final ScheduledExecutorService executor;
+  private final ApiClock clock;
+  private final Duration waitTimeout;
+  private final Duration idleTimeout;
+
+  private ScheduledFuture<?> future;
+
+
+  public WatchdogInterceptor(Set<MethodDescriptor<?, ?>> watchedMethods,
+      ScheduledExecutorService executor,
+      ApiClock clock, Duration waitTimeout, Duration idleTimeout) {
```

Comment:
We don't use `Duration`s or `ApiClock` anywhere else.  Can you please raise a bug to track unification of these concepts?

Reply:
done